### PR TITLE
build: replace setup.py with pyproject.toml

### PR DESCRIPTION
**Pull request by baggiponte**
Addresses #54. The library builds correctly; I also ran the tests and I get some failures because some resources from nltk are not installed, e.g.:

```
>       raise LookupError(resource_not_found)
E       LookupError:
E       **********************************************************************
E         Resource punkt not found.
E         Please use the NLTK Downloader to obtain the resource:
E
E         >>> import nltk
E         >>> nltk.download('punkt')
E
E         For more information see: https://www.nltk.org/data.html
E
E         Attempted to load tokenizers/punkt/PY3/english.pickle
E
E         Searched in:
E           - '/Users/luca/nltk_data'
E           - '/Users/luca/dev/contrib/datatrove/venv/nltk_data'
E           - '/Users/luca/dev/contrib/datatrove/venv/share/nltk_data'
E           - '/Users/luca/dev/contrib/datatrove/venv/lib/nltk_data'
E           - '/usr/share/nltk_data'
E           - '/usr/local/share/nltk_data'
E           - '/usr/lib/nltk_data'
E           - '/usr/local/lib/nltk_data'
E           - ''
E       **********************************************************************
```

I have a fix for this but I think it should be another PR.

Notes: the following `setup.py` sections are not mapped to their equivalent in `pyproject.toml` because setuptools has package discovery:

```
setup(
    ...
    packages=find_packages("src"),
    package_dir={"": "src"},
)
```

While `include_package_data=True,` is true by default.

One thing: the `project.scripts` you create wil be available under the names `check_dataset` etc (this is regardless of whether you want to use pyproject or not). If you want them as subcommands of a `datatrove` lib, you should use `click` or `typer` to write a CLI.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,16 +1,120 @@
...
+datatrove = ["assets/*"]
+
 [tool.ruff]
-# Ignored rules:
-#   "E501" -> line length violation
-#   "C901" -> `function_name` is too complex
-ignore = ["C901", "E501"]
-select = ["C", "E", "F", "I", "W"]
+ignore = [
+  "C901", # line length violation
```
Comment (guipenedo) [inline pyproject.toml:98]: the comments ended up switched I believe. C901 is "function name is too complex"

**Comment by guipenedo** [comment]
Tests seem to work (you need to run `python -m nltk.downloader punkt` before running the tests)

**Comment by baggiponte** [comment]
> Tests seem to work (you need to run `python -m nltk.downloader punkt` before running the tests). We just have to update the code quality action as that currently checks setup.py

yes! I would suggest/implement adding a try-catch blog that does that for you.

File: pyproject.toml [hunk 2]
```diff
@@ -1,16 +1,120 @@
+[build-system]
+requires = ["setuptools"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "datatrove"
+version = "1.0.1.dev0"  # expected format is one of x.y.z.dev0, or x.y.z.rc1 or x.y.z (no to dashes, yes to dots)
+description = "HuggingFace library to process and filter large amounts of webdata"
```
Comment (guipenedo) [inline pyproject.toml:8]: is it not possible to also include the `long_description` (the readme) as before?

**Review by guipenedo — CHANGES_REQUESTED** [review]
Could you also remove `setup.py` from `.github/workflows/ci.yml`'s ruff commands?

File: pyproject.toml (hunk 1 above)
Comment (baggiponte) [reply pyproject.toml]: Solved in the last commit I pushed, thanks for spotting that :)

File: pyproject.toml (hunk 2 above)
Comment (baggiponte) [reply pyproject.toml:8]: The pyproject [spec](https://packaging.python.org/en/latest/specifications/pyproject-toml/) follows PEP621 and does not include a `long_description` field, but the readme can be included and specified in the `readme` key of the `project` table. Should be line 9 in this doc.

**Review by guipenedo — APPROVED** [review]

**Pull request merged**